# Getting published data on Della 

To run snpArcher you need sequencing data and a reference genome to map it to. 

One of Della's (many) interesting quirks is that is doesn't have internet access on the compute nodes. What this means for us is that, even though snpArcher has the capability to pull data and a reference genome from e.g. NCBI for you, it won't be able to. Instead, all of your data--reference genome, generated sequencing data and any published sequencing data you might want to use--must be downloaded locally to Della before you can run snpArcher. 

## (1) Downloading a reference genome

I have a directory in my Della /scratch called `reference_genomes` where I store, well, reference genomes. To find the reference genome for your species, you can go to the [NCBI Genomes resource](https://www.ncbi.nlm.nih.gov/home/genomes/) and type in the scientific name of your species in the search bar. Or, a quick google of "species name genome ncbi" usually does the trick. 

If there's a specific reference you're interested in, it might be hosted elsewhere. This guide is for NCBI genomes, but can likely be adapted for genomes hosted elsewhere. Or, follow download instructions given by the repository or group that has the genome you want.

I download reference genomes using ftp:

1. On the NCBI page for your species look for a genome with a "reference" tag and/or a green checkmark. There may be other options besides the genome NCBI considers the current reference. These might be older genome assemblies created with older technology or newer assemblies that haven't yet been fully vetted/processed. 
    
2. Click on the link for the assembly you want to download, then click on "FTP" at the top of the page. 

3. Find the link ending in .fna.gz and copy this link (right click and "copy link")

    - This the ftp (file transfer protocol) for your reference genome sequence. 

    - The other stuff here might link to an annotation file, QC files for the assembly, coding sequences files, RNA transcript sequences, etc. You might need these later for other analyses, but we don't need them for snpArcher.

4. Once you have the ftp link to the reference genome you want to download copied, navigate to the directory where you're going to put your reference genome on Della. 

    `cd /scratch/gpfs/ml9889/reference_genomes/genomes`

5. Use `wget` to pull this genome using the ftp link. Here I'm pulling the genome for one of my favorite animals, the Pallas Cat (_Otocolobus manul_). 

    `wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/028/564/725/GCA_028564725.2_OtoMan_p1.0/GCA_028564725.2_OtoMan_p1.0_genomic.fna.gz`

6. Tools like GATK require an unzipped reference genome, so unzip the one you just downloaded.

    `gunzip GCA_028564725.2_OtoMan_p1.0_genomic.fna.gz`

There are other ways to download reference genomes. NCBI has a datasets tool -- this allows you to pull several different files all at once if you maybe wanted to pull the genomic sequence as well as the annotation file and RNA transcripts, etc. All fine -- get your genome from where ever you can, however you want, just make sure it's on Della so snpArcher can find it.

## (2) Pulling data from a sequence repository

You may have sequencing data you've generated that you want to map to this reference genome. You'll get instructions for downloading your sequencing data from whatever sequencing core sequenced your samples. 

However, you may also want to find some sequencing data generated by other people and published in some sequencing repository like NCBI, ENA, etc. More data can be helpful, because it increases the power of the mapping and calling algorithms that snpArcher uses.

There are several different approaches you might take for looking for published data. One approach is to just search on NCBI for published sequences in the Sequence Read Archive (SRA). If there is a lot of published data for your species, it might be easier to sort through by clicking on the "Send to Run Selector" link at the top of the SRA page for the species of interest. 

On Run Selector you can use the check boxes to select the specific runs you want to download. There may be several per sample or experiment, if some individuals were sequenced multiple times or different libraries were prepared. 

Once you have some data you want to download, you can click the "Accession List" option under "Selected Data". This will give you a text file with a list of accession numbers for the runs you selected. 

From here you'll need to use the [SRA-toolkit](https://github.com/ncbi/sra-tools/wiki) to download the data. This is because NCBI stores sequencing data in this special SRA format. If the data you want are somewhere else, you might be able to use Globus to transfer data from this source to Della.

I store my sequencing data in another subdirectory called `raw_seq_data`, organized into further subdirectories by species or project.